app/workers/email_archive_worker.rb

```diff
@@ -0,0 +1,85 @@
+class EmailArchiveWorker
+  include Sidekiq::Worker
+
+  sidekiq_options queue: :cleanup
+
+  LOCK_NAME = "email_archive_worker".freeze
+
+  def perform
+    Email.with_advisory_lock(LOCK_NAME, timeout_seconds: 0) do
+      start_time = Time.zone.now
+      archived = 0
+
+      EmailArchiveQuery.call.in_batches do |batch|
+        Email.transaction do
+          archived += archive_batch(batch)
+        end
+      end
+
+      log_complete(archived, start_time, Time.zone.now)
+    end
+  end
+
+private
+
+  def archive_batch(batch)
+    archived_at = Time.zone.now
+
+    import_batch(batch, archived_at)
+    Email.where(id: batch.pluck(:id)).update_all(archived_at: archived_at)
+  end
+
+  def import_batch(batch, archived_at)
+    to_import = batch.as_json.map { |e| build_email_archive(e, archived_at) }
+    columns = to_import.first.keys.map(&:to_s)
+    values = to_import.map(&:values)
+    EmailArchive.import(columns, values, validate: false)
+  end
+
+  def build_email_archive(email_data, archived_at)
+    content_change = build_content_change(email_data)
+
+    {
+      archived_at: archived_at,
+      content_change: content_change,
+      created_at: email_data["created_at"],
```

Comment:
Is it worth using `.fetch` for a lot of these methods to error out early?

Reply:
Another good idea 👍